### PR TITLE
Sass 2953

### PR DESCRIPTION
**Pull request by salman-a-dev**
### Description
Implemented Controller, View, Intergration tests for amount not surcharge page.

Add a link to the relevant story in Jira
[SASS-2953](https://jira.tools.tax.service.gov.uk/browse/SASS-2953)

##### Before Reviewing
- [x]  Have you pulled the branch down?
- [x]  Have you assigned yourself to the PR?
- [x]  Have you moved the task to “in review” on JIRA?
- [x]  Have you checked to ensure all dependencies are up to date?
- [x]  Have you checked to ensure its been rebased against the current version of main?

##### Whilst Reviewing
- [x]  Have you run the tests?
- [x]  Have you run the journey tests?
- [x]  Have you looked at the JIRA story to make sure all Acceptance Criteria has been met?

##### After Reviewing
- [ ]  Have you checked for merge conflicts or any changes in the current main that may affect the current pull request? i.e. does it need another rebase?
- [ ]  Have you checked to make sure there are no builds in the pipeline before you merge?
- [ ]  Have you moved the task to “in pipeline” on Jira?

##### Before creating PR
- [ ]  Have you run the tests?
- [ ]  Have you run the journey tests? (where applicable)
- [ ]  Have you addressed warnings where appropriate?
- [ ]  Have you rebased against the current version of main?
- [ ]  Have you checked code coverage isn’t lower than previously?

##### After PRs been raised
- [ ]  Have you checked the PR Builder passes?

### CODE REVIEW

File: it/controllers/pensions/unauthorisedPayments/NoSurchargeAmountControllerISpec.scala [hunk 1]
```diff
@@ -0,0 +1,315 @@
...
+    val poundPrefixSelector = ".govuk-input__prefix"
+    val inputSelector = "#amount"
+    val expectedErrorHref = "#amount"
+
+    def paragraphSelector(index: Int): String = s"#main-content > div > div > p:nth-of-type($index)"
+  }
+
+  object CommonExpectedEN extends CommonExpectedResults {
+    val expectedCaption: Int => String = (taxYear: Int) => s"Unauthorised payments from pensions for 6 April ${taxYear - 1} to 5 April $taxYear"
+    val expectedTitle = "Total amount that did not result in a surcharge"
```
[the remark this answers is not in the document]
Comment (salman-a-dev) [reply it/controllers/pensions/unauthorisedPayments/NoSurchargeAmountControllerISpec.scala:75]: It should be amount, I've changed this.

File: app/controllers/pensions/unauthorisedPayments/NoSurchargeAmountController.scala [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+          case Right(Some(data)) if data.pensions.unauthorisedPayments.noSurchargeQuestion.contains(true) =>
+            val pensionsCYAModel: PensionsCYAModel = data.pensions
+            val viewModel = pensionsCYAModel.unauthorisedPayments
+            val updatedCyaModel: PensionsCYAModel = pensionsCYAModel.copy(unauthorisedPayments = viewModel.copy(noSurchargeAmount = Some(amount)))
+
+            pensionSessionService.createOrUpdateSessionData(request.user,
+              updatedCyaModel, taxYear, data.isPriorSubmission)(errorHandler.internalServerError()) {
+
+              // TODO redirect to non uk tax amount no surcharge page
+              Redirect(PensionsSummaryController.show(taxYear))
```
[the remark this answers is not in the document]
Comment (salman-a-dev) [reply app/controllers/pensions/unauthorisedPayments/NoSurchargeAmountController.scala:74]: Redirecting to this page.

**Comment by erasmos-hmrc** [comment]
As @OlufemiAdeOlusile has been absent for two days, I'll become the assigned reviewer.

**Comment by erasmos-hmrc** [comment]
> Have you checked to ensure all dependencies are up to date?

They are not; however, I've decided that this will be soon addressed in [this ticket](https://jira.tools.tax.service.gov.uk/browse/SASS-3235).

**Comment by erasmos-hmrc** [comment]
> Have you checked to ensure its been rebased against the current version of main?

Yes.

File: app/controllers/pensions/unauthorisedPayments/SurchargeAmountController.scala [hunk 3]
```diff
@@ -44,7 +45,7 @@ class SurchargeAmountController @Inject()(authAction: AuthorisedAction,
 
   val amountForm: Form[BigDecimal] = AmountForm.amountForm(
     emptyFieldKey = "unauthorisedPayments.surchargeAmount.error.noEntry",
-    wrongFormatKey = "unauthorisedPayments.surchargeAmount.error.incorrectFormat"
+    wrongFormatKey = "common.error.incorrectFormat"
```
Comment (erasmos-hmrc) [inline app/controllers/pensions/unauthorisedPayments/SurchargeAmountController.scala:48]: Why was this changed? Personally, I'd rather have "scoped" messages even when duplicated. And it seems to be the more consistent pattern in this application.

File: app/controllers/pensions/unauthorisedPayments/DidYouPayNonUkTaxController.scala [hunk 4]
```diff
@@ -94,6 +94,7 @@ class DidYouPayNonUkTaxController @Inject()(implicit val cc: MessagesControllerC
                     )
                   )
               }
+              //Todo check data.pensions.unauthorisedPayments.noSurchargeQuestion redirect to NoSurchargeAmountController.show(taxYear) see SASS-3228
```
Comment (erasmos-hmrc) [inline app/controllers/pensions/unauthorisedPayments/DidYouPayNonUkTaxController.scala:97]: Always good to have the ticket.

**Review by erasmos-hmrc — COMMENTED** [review]
Minor points; suggestions mostly - though I'm a bit concerned about the common message.

**Comment by erasmos-hmrc** [comment]
I was initially surprised of all the diffs in spacing - but then I realised that many of the files weren't formatted.

Could you please do so, @salman-a-dev ?

File: app/controllers/pensions/unauthorisedPayments/SurchargeAmountController.scala (hunk 3 above)
Comment (erasmos-hmrc) [reply app/controllers/pensions/unauthorisedPayments/SurchargeAmountController.scala:48]: @salman-a-dev explained to me that there's an issue having duplicate messages (note: not message keys).

Please see [this spike](https://erasmos-solutions-limited.atlassian.net/browse/SUB-126).

File: it/controllers/pensions/unauthorisedPayments/SurchargeAmountControllerISpec.scala [hunk 5]
```diff
@@ -27,17 +27,18 @@ import play.api.http.HeaderNames
 import play.api.http.Status.{BAD_REQUEST, OK, SEE_OTHER}
 import play.api.libs.ws.WSResponse
 import utils.PageUrls.UnAuthorisedPayments.surchargeAmountUrl
+import utils.PageUrls.unauthorisedPaymentsPages.didYouPayNonUkTaxUrl
```
[the remark this answers is not in the document]
Comment (erasmos-hmrc) [reply it/controllers/pensions/unauthorisedPayments/SurchargeAmountControllerISpec.scala:30]: As agreed, we won't take care of this now.

File: it/controllers/pensions/unauthorisedPayments/NoSurchargeAmountControllerISpec.scala [hunk 6]
```diff
@@ -0,0 +1,399 @@
...
+import play.api.libs.ws.WSResponse
+import utils.PageUrls.UnAuthorisedPayments.noSurchargeAmountUrl
+import utils.PageUrls.{fullUrl, pensionSummaryUrl}
+import utils.PageUrls.unauthorisedPaymentsPages.{nonUKTaxOnAmountSurcharge, whereAnyOfTheUnauthorisedPaymentsUrl}
+import utils.{IntegrationTest, PensionsDatabaseHelper, ViewHelpers}
+
+class NoSurchargeAmountControllerISpec extends IntegrationTest with ViewHelpers with BeforeAndAfterEach with PensionsDatabaseHelper {
+
+
+  val userScenarios: Seq[UserScenario[CommonExpectedResults, SpecificExpectedResults]] = Seq(UserScenario(isWelsh = false, isAgent = false, CommonExpectedEN, None), UserScenario(isWelsh = false, isAgent = true, CommonExpectedEN, None), UserScenario(isWelsh = true, isAgent = false, CommonExpectedCY, None), UserScenario(isWelsh = true, isAgent = true, CommonExpectedCY, None))
```
[the remark this answers is not in the document]
Comment (erasmos-hmrc) [reply it/controllers/pensions/unauthorisedPayments/NoSurchargeAmountControllerISpec.scala]: Confirmed.

File: it/controllers/pensions/unauthorisedPayments/NoSurchargeAmountControllerISpec.scala [hunk 7]
```diff
@@ -0,0 +1,399 @@
...
+      import user.commonExpectedResults._
+
+      s"language is ${welshTest(user.isWelsh)} and request is from an ${agentTest(user.isAgent)}" should {
+      "render unauthorised payments that resulted in no surcharges page with no pre filling" which {
+        lazy val result: WSResponse = {
+          dropPensionsDB()
+          authoriseAgentOrIndividual(user.isAgent)
+          val viewModel = anUnauthorisedPaymentsViewModel.copy(noSurchargeAmount = None)
+          insertCyaData(pensionsUserDataWithUnauthorisedPayments(viewModel, isPriorSubmission = false), aUserRequest)
+          urlGet(fullUrl(noSurchargeAmountUrl(taxYearEOY)), user.isWelsh, follow = false, headers = Seq(HeaderNames.COOKIE -> playSessionCookies(taxYearEOY, validTaxYearList)))
```
[the remark this answers is not in the document]
Comment (erasmos-hmrc) [reply it/controllers/pensions/unauthorisedPayments/NoSurchargeAmountControllerISpec.scala]: Confirmed.

File: it/controllers/pensions/unauthorisedPayments/NoSurchargeAmountControllerISpec.scala [hunk 8]
```diff
@@ -0,0 +1,399 @@
...
+    "redirect to the non uk tax not surcharge page if user tries to access link directly" which {
+      val validAmount = "100.22"
+      val validForm: Map[String, String] = Map(AmountForm.amount -> validAmount)
+      implicit lazy val result: WSResponse = {
+        dropPensionsDB()
+
+        val viewModel = anUnauthorisedPaymentsViewModel.copy(noSurchargeQuestion = Some(false))
+        insertCyaData(pensionsUserDataWithUnauthorisedPayments(viewModel), aUserRequest)
+        authoriseAgentOrIndividual(isAgent = false)
+        urlPost(fullUrl(noSurchargeAmountUrl(taxYearEOY)), body = validForm, follow = false, headers = Seq(HeaderNames.COOKIE -> playSessionCookies(taxYearEOY, validTaxYearList)))
```
[the remark this answers is not in the document]
Comment (erasmos-hmrc) [reply it/controllers/pensions/unauthorisedPayments/NoSurchargeAmountControllerISpec.scala]: Confirmed.

**Comment by erasmos-hmrc** [comment]
I confirmed that all was reformatted.

**Comment by erasmos-hmrc** [comment]
@salman-a-dev . All resolved. I'll just run the journey tests.

**Review by erasmos-hmrc — DISMISSED** [review]

**Review by erasmos-hmrc — DISMISSED** [review]

**Review by erasmos-hmrc — APPROVED** [review]
